### PR TITLE
Generate instantiations

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Add a macro for generating many explicit instantiations from a Cartesian product of macro-tuples. This simplifies generation of explicit instantiations for function and class templates with many template parameters, or many values for each template parameter. Here by many I mean more than 2.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."

### CODE REVIEW

File: src/DataStructures/Index.cpp [hunk 1]
```diff
@@ -3,29 +3,31 @@
...
-template bool operator==(const Index<1>& lhs, const Index<1>& rhs);
-template bool operator==(const Index<2>& lhs, const Index<2>& rhs);
-template bool operator==(const Index<3>& lhs, const Index<3>& rhs);
+GENERATE_INSTANTIATIONS(INSTANTIATE, (0, 1, 2, 3))
 
-template bool operator!=(const Index<0>& lhs, const Index<0>& rhs);
-template bool operator!=(const Index<1>& lhs, const Index<1>& rhs);
-template bool operator!=(const Index<2>& lhs, const Index<2>& rhs);
-template bool operator!=(const Index<3>& lhs, const Index<3>& rhs);
+#undef GET_DIM
```
Comment (kidder) [inline src/DataStructures/Index.cpp]: is this supposed to be DIM ?

File: src/Utilities/GenerateInstantiations.hpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  BOOST_PP_TUPLE_TO_LIST(BOOST_PP_TUPLE_SIZE(elem), elem)
+/// \endcond
+
+/*!
+ * \ingroup Utilities
+ * \brief Macro useful for generating many explicit instantiations of function
+ * or class templates
+ *
+ * It is often necessary to generate explicit instantiations of function or
+ * class templates. Since the total number of explicit instantiation scales the
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: scales as the

File: src/Utilities/GenerateInstantiations.hpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+ * product of the number of possible number of parameter values of each template
+ * parameter, this quickly becomes tedious. This macro allows you to easily
+ * generate hundreds of explicit instantiations.
+ *
+ * The first argument to the macro is a macro that takes two argument described
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: two arguments

File: src/Utilities/GenerateInstantiations.hpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ * below. The remaining arguments are macro-tuples, e.g. `(1, 2, 3)`. The
+ * Cartesian product of the macro-tuples is then computed and each term is
+ * passed as a tuple as the second argument to the `INSTANTIATION_MACRO`. The
+ * first argument to the `INSTANTIATION_MACRO` in a Boost.Preprocessor internal
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: this sentence should be by the macro that uses the `_` as I find it to be confusing here

File: src/Utilities/GenerateInstantiations.hpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ * GENERATE_INSTANTIATIONS(INSTANTIATE, (1, 2, 3), (double, DataVector))
+ *
+ * #undef DIM
+ * #undef DTYPE
+ * #undef INSTANTIATE
+ * \endcode
+ *
+ * We don't show the result from preprocessor since this would be roughly 150
+ * lines, but you can hopefully see the benefits of generating explicit
+ * instantiations using `GENERATE_INSTANTIATIONS` way.
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: using the

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Updated! Let me know if the sentence about the `_` is clearer now.

File: src/Utilities/GenerateInstantiations.hpp [hunk 6]
```diff
@@ -28,26 +28,27 @@
...
- * The first argument to the macro is a macro that takes two argument described
- * below. The remaining arguments are macro-tuples, e.g. `(1, 2, 3)`. The
- * Cartesian product of the macro-tuples is then computed and each term is
+ * class templates. Since the total number of explicit instantiation scales as
+ * the product of the number of possible number of parameter values of each
+ * template parameter, this quickly becomes tedious. This macro allows you to
+ * easily generate hundreds of explicit instantiations.
+ *
+ * The first argument to the macro is a macro that takes two arguments and
+ * described below. The remaining arguments are macro-tuples, e.g. `(1, 2, 3)`.
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: is described

File: src/Utilities/GenerateInstantiations.hpp [hunk 7]
```diff
@@ -28,26 +28,27 @@
...
+ * The Cartesian product of the macro-tuples is then computed and each term is
  * passed as a tuple as the second argument to the `INSTANTIATION_MACRO`. The
- * first argument to the `INSTANTIATION_MACRO` in a Boost.Preprocessor internal
- * variable so just make it `_`. For example, to generate explicit
- * instantiations of the class `Index<Dim>` for `Dim = 0,1,2,3` you would write:
+ * first argument to the `INSTANTIATION_MACRO` is a Boost.Preprocessor internal
+ * variable so just make it `_`. See the `INSTANTIATION(_, data)` macro below
+ * serves as example. A concrete example is generating explicit instantiations
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: serves as -> as an

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Okay, another update. Let me know if the documentation can be improved in any other way :)

**Review by kidder — DISMISSED** [review]

**Review by kidder — CHANGES_REQUESTED** [review]
you have a doxygen warning to deal with

File: src/Executables/DebugPreprocessor/DebugPreprocessor.cpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Utilities/GenerateInstantiations.hpp"
+
+#define GET_DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
```
Comment (kidder) [inline src/Executables/DebugPreprocessor/DebugPreprocessor.cpp]: // This is an example:  replace the code with what you want to debug

File: src/Utilities/GenerateInstantiations.hpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  BOOST_PP_TUPLE_TO_LIST(BOOST_PP_TUPLE_SIZE(elem), elem)
+/// \endcond
+
+/*!
+ * \ingroup Utilities
+ * \brief Macro useful for generating many explicit instantiations of function
+ * or class templates
+ *
+ * It is often necessary to generate explicit instantiations of function or
+ * class templates. Since the total number of explicit instantiation scales as
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: instantiations

File: src/Utilities/GenerateInstantiations.hpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ * #define INSTANTIATE(_, data)                                              \
+ *   template Scalar<DTYPE(data)>                                            \
+ *   ScalarWave::Solutions::PlaneWave<DIM(data)>::psi(                       \
+ *       const tnsr::I<DTYPE(data), DIM(data)>& x, double t) const noexcept; \
+ *   template Scalar<DTYPE(data)>                                            \
+ *   ScalarWave::Solutions::PlaneWave<DIM(data)>::dpsi_dt(                   \
+ *       const tnsr::I<DTYPE(data), DIM(data)>& x, double t) const noexcept; \
+ *   template tnsr::i<DTYPE(data), DIM(data)>                                \
+ *   ScalarWave::Solutions::PlaneWave<DIM(data)>::dpsi_dx(                   \
```
Comment (kidder) [inline src/Utilities/GenerateInstantiations.hpp]: remove the rest

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**